File: lib-mvpbase-appcompat/src/main/java/com/baozi/mvp/presenter/PagerFragmentPresenter.java
package com.baozi.mvp.presenter;

import android.content.res.ColorStateList;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.baozi.mvp.view.PagerFragmentView;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Administrator on 2017/8/21 0021.
 */

public class PagerFragmentPresenter {
    private List<Fragment> fragments;
    private PagerFragmentView mView;

    private static Fragment getFragment(Class<? extends Fragment> zClass) {
        try {
            return zClass.newInstance();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        return null;
    }

    public PagerFragmentPresenter(PagerFragmentView mView) {
        this.mView = mView;
    }

    public void onCreate() {
        initViewpager();
        initTabLayout();
    }

    private void initTabLayout() {
        TabLayout tablayout = mView.getTablayout();
        if (tablayout == null) {
            return;
        }
        tablayout.setupWithViewPager(mView.getViewPager());
        //条目数
        int size = getFragments().size();

//        int tabLayoutItem = mView.getTabLayoutItem();
        String[] tabString = mView.getTabString();
        int[] tabImage = mView.getTabDrawables();
        for (int i = 0; i < size; i++) {
            TabLayout.Tab tab = tablayout.getTabAt(i);
            int image = 0;
            if (tabImage != null) {
                image = tabImage[i];
            }
            String tabStr = "";
            if (tabString != null) {
                tabStr = tabString[i];
            }
            bindTab(tab, image, tabStr);
        }
        mView.getViewPager().setOffscreenPageLimit(mView.getFragments().length);
        if (!mView.isAnimation()) {
            tablayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
                @Override
                public void onTabSelected(TabLayout.Tab tab) {
                    int position = tab.getPosition();
                    mView.getViewPager().setCurrentItem(position, false);
                }

                @Override
                public void onTabUnselected(TabLayout.Tab tab) {

                }

                @Override
                public void onTabReselected(TabLayout.Tab tab) {

                }
            });
        }
    }

    private void bindTab(TabLayout.Tab tab, int image, String tabStr) {
        int tabLayoutItem = mView.getTabLayoutItem();
        if (tabLayoutItem == 0 && tab != null) {
            tab.setText(tabStr);
            if (image == 0) {
                return;
            }
            tab.setIcon(image);
            return;
        }
        if (tabLayoutItem != 0) {
            ViewGroup inflate = (ViewGroup) LayoutInflater.from(mView.getContext()).inflate(tabLayoutItem, null);
            int childCount = inflate.getChildCount();
            for (int j = 0; j < childCount; j++) {
                View childAt = inflate.getChildAt(j);
                if (childAt instanceof ImageView) {
                    if (image != 0) {
                        childAt.setVisibility(View.VISIBLE);
                        ((ImageView) childAt).setImageResource(image);
                    } else {
                        childAt.setVisibility(View.GONE);
                    }
                }
                if (childAt instanceof TextView) {
                    childAt.setVisibility(View.VISIBLE);
                    ColorStateList tabTextColors = mView.getTablayout().getTabTextColors();
                    ((TextView) childAt).setText(tabStr);
                    ((TextView) childAt).setTextColor(tabTextColors);
                }
            }
            tab.setCustomView(inflate);
        }
    }

    private void initViewpager() {
        ViewPager viewPager = mView.getViewPager();
        viewPager.setAdapter(getAdapter());
    }

    public List<Fragment> getFragments() {
        if (fragments == null) {
            fragments = new ArrayList<>();
            Class<? extends Fragment>[] fragments = mView.getFragments();
            for (int i = 0; i < fragments.length; i++) {
                Fragment fragment = getFragment(fragments[i]);
                this.fragments.add(fragment);
            }
        }
        return fragments;
    }

    protected FragmentStatePagerAdapter getAdapter() {
        return new FragmentStatePagerAdapter(mView.getFgManager()) {
            @Override
            public Fragment getItem(int position) {
                if (getFragments().get(position) == null) {
                    Fragment fragment = getFragment(mView.getFragments()[position]);
                    getFragments().set(position, fragment);
                }
                return getFragments().get(position);
            }

            @Override
            public CharSequence getPageTitle(int position) {
                return mView.getTabString()[position];
            }

            @Override
            public int getCount() {
                return getFragments().size();
            }
        };
    }
}

File: lib-mvpbase-kotlin/src/main/java/com/linfeng/mvp/presenter/PagerFragmentPresenter.kt
package com.linfeng.mvp.presenter

import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentStatePagerAdapter
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView


import java.util.ArrayList

/**
 * Created by Administrator on 2017/8/21 0021.
 */

class PagerFragmentPresenter(private val mView: com.linfeng.mvp.view.PagerFragmentView) : PagerPresenter(mView) {
    private var fragments: ArrayList<Fragment> = ArrayList()

    override val adapter: FragmentStatePagerAdapter
        get() = object : FragmentStatePagerAdapter(mView.fgManager) {
            override fun getItem(position: Int): Fragment {
                val fragment = getFragment(mView.fragments[position])
                getFragments()[position] = fragment!!
                return getFragments()[position]
            }

            override fun getPageTitle(position: Int): CharSequence? {
                return mView.tabString?.get(position)
            }

            override fun getCount(): Int {
                return getFragments().size
            }
        }

    override fun onCreate() {
        initViewpager()
        initTabLayout()
    }

    private fun initTabLayout() {
        val tabLayout = mView.tablayout
        if (tabLayout == null || mView.tabDrawables == null || mView.tabString == null) {
            return
        }
        tabLayout.setupWithViewPager(mView.viewPager)

        val tabImage = mView.tabDrawables ?: IntArray(0)
        val tabString = mView.tabString
        val tabLayoutItem = mView.tabLayoutItem ?: 0
        for (i in tabImage.indices) {
            var tab: TabLayout.Tab? = tabLayout.getTabAt(i)
            if (tab != null && tabLayoutItem != 0) {
                val inflate = LayoutInflater.from(mView.mContext).inflate(tabLayoutItem, null) as ViewGroup
                val childCount = inflate.childCount
                for (j in 0 until childCount) {
                    val childAt = inflate.getChildAt(j)
                    if (childAt is ImageView) {
                        childAt.setImageResource(tabImage[i])
                    }
                    if (childAt is TextView) {
                        childAt.text = tabString!![i]
                    }
                }
                tab.customView = inflate
            } else {
                tab = tabLayout.newTab()
                tab.text = tabString!![i]
                tab.setIcon(tabImage[i])
                tabLayout!!.addTab(tab, i)
            }
        }
        mView.viewPager.offscreenPageLimit = mView.fragments.size
        if (!mView.isAnimation) {
            tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
                override fun onTabSelected(tab: TabLayout.Tab) {
                    val position = tab.position
                    mView.viewPager.setCurrentItem(position, false)
                }

                override fun onTabUnselected(tab: TabLayout.Tab) {

                }

                override fun onTabReselected(tab: TabLayout.Tab) {

                }
            })
        }
    }

    private fun initViewpager() {
        val viewPager = mView.viewPager
        viewPager.adapter = adapter
    }

    fun getFragments(): ArrayList<Fragment> {
        val fragments = mView.fragments
        for (i in fragments.indices) {
            val fragment = getFragment(fragments[i]) ?: continue
            this.fragments.add(fragment)
        }
        return this.fragments
    }

    companion object {
        fun getFragment(zClass: Class<Fragment>): Fragment? {
            try {
                return zClass.newInstance()
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }
            return null
        }
    }
}
